File: bindings/java/src/main/java/systems/zlink/contracts/service/discovery/Discovery.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.service.discovery;

import systems.zlink.contracts.core.RoutingId;
import systems.zlink.contracts.service.registry.MemberPeerEntry;
import systems.zlink.contracts.service.spot.ActorRoute;

import java.util.List;

/**
 * Fixed-channel discovery view contract.
 *
 * <p>One instance tracks exactly one auto-connect type and channel name, then
 * exposes spot owner resolution plus member peer snapshots for that view.
 */
public interface Discovery extends AutoCloseable {
    int routeValueMaxSize();

    void connectRegistry(String registryEndpoint);

    SpotRoute resolveSpot(RoutingId spotRid);

    ActorRoute resolveActor(String actorId);

    void bindRoute(int kind, byte[] key, byte[] value);

    void unbindRoute(int kind, byte[] key);

    DiscoveryRoute resolveRoute(int kind, byte[] key);

    void setValue(long value);

    long getValue();

    void setSpotOwnerSyncEnabled(boolean enabled);

    boolean isSpotOwnerSyncEnabled();

    void setActorRouteSyncEnabled(boolean enabled);

    boolean isActorRouteSyncEnabled();

    void setTlsClient(String caCertPem, String hostname,
                                      boolean trustSystem);

    List<MemberPeerEntry> memberPeers();

    @Override
    void close();
}

File: bindings/python/src/zlink/_runtime/service/discovery.py
# SPDX-License-Identifier: MPL-2.0
#
# Re-export shim. The discovery service implementation lives in the public
# contract source at zlink/contracts/service/discovery.py.

from ...contracts.service.discovery import *  # noqa: F401,F403


def __getattr__(name):
    from ...contracts.service import discovery as _impl

    try:
        return getattr(_impl, name)
    except AttributeError:
        raise AttributeError(f"module {__name__!r} has no attribute {name!r}") from None
